src/main/java/net/glowstone/ConsoleManager.java

```diff
@@ -296,46 +270,57 @@ public synchronized void flush() {
     private class CommandCompleter implements Completer {
 
         @Override
-        public int complete(String buffer, int cursor, List<CharSequence> candidates) {
+        public void complete(LineReader reader, ParsedLine line, List<Candidate> candidates) {
+            List<String> completions = null;
             try {
-                List<String> completions = server.getScheduler()
-                        .syncIfNeeded(() -> server.getCommandMap().tabComplete(sender, buffer));
-                if (completions == null) {
-                    return cursor;  // no completions
-                }
-                candidates.addAll(completions);
+                completions = server.getScheduler().syncIfNeeded(() -> server.getCommandMap()
+                    .tabComplete(sender, line.line()));
+            } catch (Exception e) {
+                LocalizedStrings.Console.Error.Manager.TAB_COMPLETE.log(e);
+            }
 
-                // location to position the cursor at (before autofilling takes place)
-                return buffer.lastIndexOf(' ') + 1;
-            } catch (Throwable t) {
-                logger.log(Level.WARNING, "Error while tab completing", t);
-                return cursor;
+            if (completions != null) {
+                completions.forEach(completion -> candidates.add(new Candidate(completion)));
             }
         }
     }
 
     private class ConsoleCommandThread extends Thread {
 
+        ConsoleCommandThread() {
+            setName("ConsoleCommandThread"); // NON-NLS
+            setDaemon(true);
+        }
+
         @Override
         public void run() {
-            String command = "";
+            String command = null;
             while (running) {
                 try {
-                    if (jline) {
-                        command = reader.readLine(CONSOLE_PROMPT, null);
+                    if (color) {
+                        command = reader.readLine(colorize(ChatColor.RESET.toString())
+                            + CONSOLE_PROMPT);
                     } else {
-                        command = reader.readLine();
+                        command = reader.readLine(CONSOLE_PROMPT);
                     }
-
-                    if (command == null || command.trim().isEmpty()) {
-                        continue;
+                    if (command != null && !(command = command.trim()).isEmpty()) {
+                        reader.getTerminal().writer().println(colorize(ChatColor.RESET
+                            + "====" + ChatColor.GOLD + "g>" + ChatColor.RESET + '"' + command
+                                + '"')); // NON-NLS
+                        if (command.startsWith("$")) {  // NON-NLS
+                            server.getScheduler().runTask(null,
+                                new EvalTask(command.substring(1), command.startsWith("$$")));
```

Comment:
Add a config setting that disables this mode. Disable it by default, unless that would break mod compatibility, and maybe even then. That way, gaining control of the console doesn't allow arbitrary code execution if the legit admin doesn't use it.